pkg/controller/buildrun/buildrun_controller.go

```diff
@@ -232,42 +244,87 @@ func (r *ReconcileBuildRun) Reconcile(request reconcile.Request) (reconcile.Resu
 	ctxlog.Debug(ctx, "starting reconciling request from a BuildRun or TaskRun event", namespace, request.Namespace, name, request.Name)
 
 	buildRun = &buildv1alpha1.BuildRun{}
-	build = &buildv1alpha1.Build{}
 	lastTaskRun := &v1beta1.TaskRun{}
 
 	// for existing TaskRuns update the BuildRun Status, if there is no TaskRun, then create one
-	if err := r.client.Get(ctx, types.NamespacedName{Name: request.Name, Namespace: request.Namespace}, lastTaskRun); err != nil && apierrors.IsNotFound(err) {
+	if err := r.client.Get(ctx, types.NamespacedName{Name: request.Name, Namespace: request.Namespace}, lastTaskRun); err != nil {
+		if apierrors.IsNotFound(err) {
+			err = r.GetBuildRunObject(ctx, request.Name, request.Namespace, buildRun)
+			if err != nil && !apierrors.IsNotFound(err) {
+				return reconcile.Result{}, err
+			} else if apierrors.IsNotFound(err) {
+				// If the BuildRun and TaskRun are not found, it might mean that we are running a Reconcile after a TaskRun was deleted. If this is the case, we need
+				// to identify from the request the BuildRun name associate to it and update the BuildRun Status.
+				r.VerifyRequestName(ctx, request, buildRun)
+				return reconcile.Result{}, nil
+			}
 
-		err = r.GetBuildRunObject(ctx, request.Name, request.Namespace, buildRun)
-		if err != nil && !apierrors.IsNotFound(err) {
-			return reconcile.Result{}, err
-		} else if apierrors.IsNotFound(err) {
-			// If the BuildRun and TaskRun are not found, it might mean that we are running a Reconcile after a TaskRun was deleted. If this is the case, we need
-			// to identify from the request the BuildRun name associate to it and update the BuildRun Status.
-			r.VerifyRequestName(ctx, request, buildRun)
-			return reconcile.Result{}, nil
-		}
+			// The following block is there to cover the scenario where the build controller is restarted for some reason. In this case it reconciles all buildruns. It then
+			// must not create a second TaskRun for a BuildRun. A good way to test this scenario is by running the controller in a cluster where Tekton is NOT running.
+			// Because then the created TaskRun is not getting processed and our BuildRun will not get the LatestTaskRunRef field set.
+			if buildRun.Status.BuildSpec != nil {
+				// we already worked on this BuildRun and potentially there is already a TaskRun
+				if buildRun.Status.LatestTaskRunRef != nil {
+					// that was easy, the TaskRun is referenced in the BuildRun
+					ctxlog.Info(ctx, "BuildRun already has TaskRun", namespace, request.Namespace, name, request.Name)
```

Comment:
I think it is fine for now, but I am not sure if we will support `Pause`, `Cancel` or `Rerun` a buildrun in future.

Reply:
Sure, we then need to revisit certain conditions also in the `UpdateFunc` predicate for the buildrun.